test/PassPlanFilePluginTest.groovy

```diff
@@ -0,0 +1,125 @@
+import org.junit.Test
+import org.junit.Before
+import org.junit.After
+import org.junit.runner.RunWith
+import de.bechte.junit.runners.context.HierarchicalContextRunner
+
+@RunWith(HierarchicalContextRunner.class)
+class PassPlanFilePluginTest {
+    @Before
+    void resetJenkinsEnv() {
+        Jenkinsfile.instance = mock(Jenkinsfile.class)
+        when(Jenkinsfile.instance.getEnv()).thenReturn([:])
+    }
+
+    private configureJenkins(Map config = [:]) {
+        Jenkinsfile.instance = mock(Jenkinsfile.class)
+        when(Jenkinsfile.instance.getEnv()).thenReturn(config.env ?: [:])
+    }
+
+    public class Init {
+        @After
+        void resetPlugins() {
+            TerraformPlanCommand.resetPlugins()
+            TerraformApplyCommand.resetPlugins()
+            TerraformEnvironmentStage.reset()
+        }
+
+        @Test
+        void modifiesTerraformEnvironmentStageCommand() {
+            PassPlanFilePlugin.init()
+
+            Collection actualPlugins = TerraformEnvironmentStage.getPlugins()
+            assertThat(actualPlugins, hasItem(instanceOf(PassPlanFilePlugin.class)))
+        }
+
+        @Test
+        void modifiesTerraformPlanCommand() {
+            PassPlanFilePlugin.init()
+
+            Collection actualPlugins = TerraformPlanCommand.getPlugins()
+            assertThat(actualPlugins, hasItem(instanceOf(PassPlanFilePlugin.class)))
+        }
+
+        @Test
+        void modifiesTerraformApplyCommand() {
+            PassPlanFilePlugin.init()
+
+            Collection actualPlugins = TerraformApplyCommand.getPlugins()
+            assertThat(actualPlugins, hasItem(instanceOf(PassPlanFilePlugin.class)))
+        }
+
+    }
+
+    public class Apply {
+
+        @Test
+        void decoratesTheTerraformEnvironmentStage()  {
+            PassPlanFilePlugin plugin = new PassPlanFilePlugin()
+            def environment = spy(new TerraformEnvironmentStage())
+            plugin.apply(environment)
+
+            verify(environment, times(1)).decorate(eq(TerraformEnvironmentStage.PLAN), any(Closure.class))
+            verify(environment, times(1)).decorate(eq(TerraformEnvironmentStage.APPLY), any(Closure.class))
+        }
+
+        @Test
+        void runsStashPlan() {
+            def expectedClosure = { -> }
+            def plugin = spy(new PassPlanFilePlugin())
+            doReturn(expectedClosure).when(plugin).stashPlan()
+            def stage = mock(TerraformEnvironmentStage.class)
+
+            plugin.apply(stage)
+
+            verify(stage).decorate(anyString(), eq(expectedClosure))
+        }
+
+        @Test
+        void runsUnstashPlan() {
+            def expectedClosure = { -> }
+            def plugin = spy(new PassPlanFilePlugin())
+            doReturn(expectedClosure).when(plugin).unstashPlan()
+            def stage = mock(TerraformEnvironmentStage.class)
+
+            plugin.apply(stage)
+
+            verify(stage).decorate(anyString(), eq(expectedClosure))
+        }
+
```

Comment:
I see the new `runsStashPlan` runsUnstashPlan` tests - thanks for adding that!

I'm looking at  the code coverage however, and it doesn't seem to have changed. Digging into the details of the test, it looks like you're mocking the closures, and the closures are the lines of code missing coverage.

The closure  is  the thing that I was interested in getting coverage on.  If there were, say, a syntactic error, simply exercising the closure would catch that.  Instead of exercising `stashPlan` and `unstashPlan`  through `apply`, could we break it down to a more fine-grained unit test, exercise those methods directly?

Something along the lines of: 
```
class UnstashPlan {
        @Test
        void justExercisingNotValidatingBehavior() {
            def plugin = new PassPlanFilePlugin()

            def unstashClosure = plugin.unstashPlan('myenv')
            unstashClosure.call { -> }
        }
}
```

^-- if you did the above, we should see code coverage go up, because we're actually exercising the closure.

If we wanted to take this further, we could even validate *some* amount of behavior.  It's really important that the inner closure is executed - if calling the inner closure were mistakenly left off, it would have really detrimental impact on the behavior of the pipeline.  So we could build more safety into our unit tests with something like:

```
class UnstashPlan {
        @Test
        void executesPassedClosure() {
            def wasCalled = false
            def innerClosure = { -> wasCalled = true }
            def plugin = new PassPlanFilePlugin()

            def unstashClosure = plugin.unstashPlan('myenv')
            unstashClosure.call(innerClosure)

            assertTrue(wasCalled)
        }
```

Reply:
@kmanning I tried implementing that second block but I'm getting some errors.

Any ideas what to do about `groovy.lang.MissingMethodException: No signature of method: PassPlanFilePlugin.echo() is applicable for argument types: (org.codehaus.groovy.runtime.GStringImpl) values: [Stashing tfplan-dev file]`

Reply:
That makes sense.  Your closure is running Jenkinsfile DSL methods, like echo.  

```
    public Closure unstashPlan(String env) {
        return { closure ->
            String planFile = "tfplan-" + env
            echo "Unstashing ${planFile} file"
            unstash planFile
            closure()
        }
```

`echo` isn't defined in  the plugin, and it's not defined in your test.  When executing a closure, when the VM comes across a method, it does a search.  See [here](https://groovy-lang.org/closures.html#_delegation_strategy):
1. `this`
2. `owner`
3. `delegate`

If the method can't be  found after traversing those objects, it throws the MissingMethodException you're seeing. 

In a "normal" terraform-pipeline, the owner or delegate would be the actual Jenkinsfile executing on Jenkins.  But you're just in a unit test.  So what you can do, is explicitly set a delegate, and use the DummyJenkinsfile to stub in for what would normally happen in Jenkins.

As an example, you can check  out this test: https://github.com/manheim/terraform-pipeline/blob/9bac14b75ce353fbbe7592e45a9ec786e5824ab9/test/TerraformEnvironmentStageTest.groovy#L168-L178